testing/gerbv_example/multivibrator-basename/expected/testbase_name_milldrill.ngc

```diff
@@ -67,7 +67,7 @@ G0 Z0.08000
 
 G0 X-3.81500 Y-2.66535
 G1 Z1.60000 F50.00000
-G1 F100.00000
+G1 F50.00000
 G1 X-3.81500 Y-2.58665 Z0.80000
```

Comment:
I don't like this that much... I think this is too slow and slow is not always better. It generates more heat on the endmill which is not good for tool life.

In this example cutter->feed is a little bit to fast and then it drops completely to cutter->vertfeed which takes twice as long (in this example). Considering zigzaging is not optimal (but the best for this situation anyway) as the cutter engagement is already less then specified (except at the very end of each line) and thus far slower with smaller chips for heat extraction. (Although I don't know if this really makes a difference with 1,6mm FR2 or FR4 material...) I don't feel this is right.

If you don't like constant horizontal feed for zigzags and helixes, I can see if I can calculate the optimal speed to be within both parameters but not fall back to vertical feed completely. Most of the work is done anyway. You calculated the horizontal and vertical distance already and we have both feeds, so we can calculate the optimal feed for the hypotenuse.

Reply:
I wrote a hypotenuse version, too, one that sets the path so that the vertical component is less than or equal to the max very speed and the horizontal component is less than or equal to the horizontal speed.

I am not a CNC expert so I don't know what the right thing to do it. But it seems to me that if the plunge speed is way less than the cut speed and the hole is just slightly larger than the bit, it should be pretty similar to the vertical speed and not the horizontal one, right?

I had it the other way before but I worried that it would look weird to have feedrates that are neither cut speed nor plunge speed but something else. I'll do it again.

Reply:
I'm no CNC expert by any means either. I agree that it is kind of weird to the feed to a calculated number. I don't think I have seen that before anywere else. But I also don't like to go too slow.

What if we look at this from a different angle? As I understand it the vertical feedrate is less than the horizontal feed because it puts more strain on the endmill that is not really designed for drilling. But when we do a helix or zigzag we relieve most of the extra strain on the tool when the angle is narrow enough and could use the horizontal feedrate even if it exceeds the set vertical feed rate. Now the only question is what angle is "narrow enough"? We could either use an arbitrary value or make it user configurable. And if the downward angle is greater than this value we use the vertical feed otherwise the horizontal feed. I think a sensible default value would be between 15° and 20°.

From https://www.cnccookbook.com/helical-interpolation-ramp-angle-and-the-best-ways-to-enter-a-cut/

> A lot of machinists are pretty conservative on their ramp angle.  I see recommendations like 1.5 degrees or 2.5 degrees all the time.  You can go faster than you’d think.  OSG talks about 10 to 20 degrees in tough materials for some of their endmills.  Granted, those are premium endmills, but don’t be afraid to be a little more aggressive.

As we really don't expect any "tough" materials I think this should be fine.

Reply:
Just realized that my "solution" would probably make it use the vertical feed even more. It would need to be a really long slot or shallow steps for it not to surpass 20°. So maybe your solution is already the best? I don't know... How do you want to solve this?

Reply:
I also don't know!  I'm trying to find a good guide online.  This page has a guideline:

https://www.harveyperformance.com/in-the-loupe/types-tool-entry/

> helix diameter of greater than 110-120% of the cutter diameter

So if `mill_radius*2 > cutter_diamter * 1.1` then use the horizontal feedrate, otherwise the vertical feedrate?  That doesn't even take the cut-infeed into account.  Should it?

If someone is using milldrill and has specified a vertical speed that is different from the horizontal speed, and the hole to bore is very nearly straight down, I think that there would be some expectation that the vertical speed is used!  Otherwise, why did the user specify it?  And a hole is "very nearly striaght down" if the path done on each pass is really small, such as being less than 110% of cutter-diameter.

I kind of like the idea of *not* doing interpolation because fewer users will be, like, "Hey, what the hell is this weird number that I didn't specify?"

Reply:
Yeah, something like that.  They talk about the ratio of hole diameter but with slots, it's a path that isn't round.  So maybe it's actually:

`mill_radius*2*pi + distance*2 > cutter_diameter*pi * 1.1`

Where distance is 0 when we're not doing a slot and mill_radius is zero when it's just a zig-zag or a plunge.

Reply:
I think it's right the way you first wrote it. I don't think we need to calculate the length of the path and compare this to the cutter, but the length we are trying to clear. It's about clearance next to the cutter (to check if the chips get evacutated more or less properly).

So it would be `mill_radius*2 + distance > cutter_diamter * 1.1` (for a distance of 0 for holes).

Reply:
Thinking about it maybe it would be better to increase 1.1 for the zigzag-case a bit. In the helix case there is 10% room next to the cutter in all directions, but for the zigzag-case it's just on two sides, so maybe increase this a bit? 20% 30%?